test/test_roo.rb

```diff
@@ -1,1967 +1,1975 @@
+#STDERR.reopen "/dev/null","w"
+
+require 'test_helper'
+
+class TestRoo < Minitest::Test
+
+  OPENOFFICE   = true   # do OpenOffice-Spreadsheet Tests? (.ods files)
+  EXCELX       = true   # do Excelx Tests? (.xlsx files)
+  LIBREOFFICE  = true   # do LibreOffice tests? (.ods files)
+  CSV          = true   # do CSV tests? (.csv files)
+
+  FORMATS = [
+    :excelx,
+    :openoffice,
+    :libreoffice
+  ]
+
+  ONLINE = false
+  LONG_RUN = false
+
+  def fixture_filename(name, format)
+    case format
+    when :excelx
+      "#{name}.xlsx"
+    when :openoffice, :libreoffice
+      "#{name}.ods"
+    else
+      raise ArgumentError, "unexpected format #{format}"
+    end
+  end
+
+  # call a block of code for each spreadsheet type
+  # and yield a reference to the roo object
+  def with_each_spreadsheet(options)
+    if options[:format]
+      formats = Array(options[:format])
+      invalid_formats = formats - FORMATS
+      unless invalid_formats.empty?
+        raise "invalid spreadsheet types: #{invalid_formats.join(', ')}"
+      end
+    else
+      formats = FORMATS
+    end
+    formats.each do |format|
+      begin
+        yield Roo::Spreadsheet.open(File.join(TESTDIR,
+          fixture_filename(options[:name], format)))
+      rescue => e
+        raise e, "#{e.message} for #{format}", e.backtrace
+      end
+    end
+  end
+
+  def test_sheets_csv
+    if CSV
+      oo = Roo::CSV.new(File.join(TESTDIR,'numbers1.csv'))
+      assert_equal ["default"], oo.sheets
+      assert_raises(RangeError) { oo.default_sheet = "no_sheet" }
+      assert_raises(TypeError)  { oo.default_sheet = [1,2,3] }
+      oo.sheets.each { |sh|
+        oo.default_sheet = sh
+        assert_equal sh, oo.default_sheet
+      }
+    end
+  end
+
+  def test_sheets
+    with_each_spreadsheet(:name=>'numbers1') do |oo|
+      assert_equal ["Tabelle1","Name of Sheet 2","Sheet3","Sheet4","Sheet5"], oo.sheets
+      assert_raises(RangeError) { oo.default_sheet = "no_sheet" }
+      assert_raises(TypeError)  { oo.default_sheet = [1,2,3] }
+      oo.sheets.each { |sh|
+        oo.default_sheet = sh
+        assert_equal sh, oo.default_sheet
+      }
+    end
+  end
+
+  def test_cells
+    with_each_spreadsheet(:name=>'numbers1') do |oo|
+      # warum ist Auswaehlen erstes sheet hier nicht
+      # mehr drin?
+      oo.default_sheet = oo.sheets.first
+      assert_equal 1, oo.cell(1,1)
+      assert_equal 2, oo.cell(1,2)
+      assert_equal 3, oo.cell(1,3)
+      assert_equal 4, oo.cell(1,4)
+      assert_equal 5, oo.cell(2,1)
+      assert_equal 6, oo.cell(2,2)
+      assert_equal 7, oo.cell(2,3)
+      assert_equal 8, oo.cell(2,4)
+      assert_equal 9, oo.cell(2,5)
+      assert_equal "test", oo.cell(2,6)
+      assert_equal :string, oo.celltype(2,6)
+      assert_equal 11, oo.cell(2,7)
+      # TODO: This has been failing for Salsify for a long time
+      # unless oo.kind_of? Roo::CSV
+      #   assert_equal :float, oo.celltype(2,7)
```

Comment:
Interesting, what are all of these if not floats? Of course I can't find any tooling in Excel that would tell us what to expect

Reply:
They're sometimes a `:string`. I don't think there's a type `:integer`. Float had bad side effects for integers in spreadsheets.